File: JavaScript/002_add_two_numbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    let result = new ListNode(0);
    let currentNode = result;
    let previousNode = null;

    while(l1 != null || l2 != null){
        if(l1 === null){
            currentNode.val += l2.val;
        } else if (l2 === null){
            currentNode.val += l1.val;
        } else {
            currentNode.val += l1.val + l2.val;
        }

        currentNode.next = new ListNode(0);

        if(currentNode.val >= 10){
            currentNode.val = currentNode.val % 10;
            currentNode.next.val += 1;
        }
        
        previousNode = currentNode;
        currentNode = currentNode.next;

        if(l1 != null){
            l1 = l1.next;
        }
        if(l2 != null){
            l2 = l2.next;
        }
    }

    if(currentNode.val === 0){
        previousNode.next = null;
    }

    return result;
};

File: Python/002_add_two_numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        shift = 0
        result = ListNode(0)
        origin_result = result

        while l1 or l2:
            v1_value = l1.val if l1 else 0
            v2_value = l2.val if l2 else 0

            result.val += v1_value + v2_value
            if result.val >= 10:
                result.val = result.val - 10
                shift = 1

            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None
            if not l1 and not l2 and not shift:
                break

            result.next = ListNode(shift)
            result = result.next
            shift = 0

        return origin_result
